runners/google-cloud-dataflow-java/worker/src/main/java/org/apache/beam/runners/dataflow/worker/fn/control/BeamFnMapTaskExecutor.java

```diff
@@ -351,7 +398,115 @@ void updateProgress() {
       }
     }
 
-    private void updateMetrics(BeamFnApi.Metrics metrics) {
+    // Keeping this as static class for this iteration. Will extract to separate file and generalize
+    // when more counter types are added.
+    // todomigryz: define counter transformer factory
+    // that can provide respective counter transformer for different type of counters.
+    // (ie RowCountCounterTranformer, MSecCounterTransformer, UserCounterTransformer, etc)
+    private static class MonitoringInfoToCounterUpdateTransformer {
+
+      private final Map<String, DataflowStepContext> transformIdMapping;
+
+      public MonitoringInfoToCounterUpdateTransformer(
+          final Map<String, DataflowStepContext> transformIdMapping) {
+        this.transformIdMapping = transformIdMapping;
+      }
+
+      // todo: search code for "beam:metrics"... and replace them with relevant enums from
+      // proto after rebasing above https://github.com/apache/beam/pull/6799 that
+      // introduces relevant proto entries.
+      final String BEAM_METRICS_USER_PREFIX = "beam:metric:user";
+
+      private CounterUpdate monitoringInfoToCounterUpdate(MonitoringInfo monitoringInfo) {
+        long value = monitoringInfo.getMetric().getCounterData().getInt64Value();
+        String urn = monitoringInfo.getUrn();
+
+        String type = monitoringInfo.getType();
+
+        // todomigryz: run MonitoringInfo through validation process.
+        // refer to https://github.com/apache/beam/pull/6799
+
+        if (urn.startsWith(BEAM_METRICS_USER_PREFIX)) {
+          if (!type.equals("beam:metrics:sum_int_64")) {
+            LOG.warn(
```

Comment:
Are there scenarios where we expect invalid metric types which should be ignored? If so, please comment when this might be the case.

If a correct implementation should always send valid data, it's best to fail rather than be resilient such that it's easier to find and fix bugs.

Reply:
I believe that writing warning is good balance here. Most of the time you want your job to process data, and use counters for monitoring.
This approach allows SDK authors to find errors in their SDK implementations on one hand. And not break user pipelines on the other hand.

@ajamato Alex, what's your opinion on this topic?

Reply:
It would be good to understand the scenarios we could reach these states and design error handling appropriately.

If the only scenario that would produce this is a bug in the SDK, then we should expose that bug as early as possible to get it fixed. Log-and-continue means that we'll ignore it, and unless we have an explicit test our users will notice before we do.

I doubt we have a consistent methodology for error handling in Beam, but I remember this came up when designing DIsplayData in the java SDK. We decided errors in constructing DisplayData from user-defined PTransforms should throw and block job submission, rather than continuing with empty Display Data. This feels similar.